### ISSUE TITLE
Add support for tracing Lambda functions

### ISSUE THREAD

**Issue by jeremy-f3d** [aws/aws-xray-sdk-ruby#20]
It'd be useful to be able to trace Lambda functions, as per the code below, but currently this doesn't work as when opening a new subsegment within a Lambda function, the SDK cannot find the current context and logs the error `can not find the current context`.

```
require 'aws-xray-sdk'

def lambda_handler(event:, context:)
    # ... some code
    XRay.recorder.begin_subsegment('subsegment_name')
    # Code to record
    XRay.recorder.end_subsegment()
    # ... some other code
end
```

**Comment by haotianw465** [comment]
Thank you for the feedback. Supporting lambda functions is in our roadmap but we don't have an ETA right now. For workaround you can use python SDK lambda support as a reference implementation: https://github.com/aws/aws-xray-sdk-python/blob/master/aws_xray_sdk/core/lambda_launcher.py. Please feel free to let me know if you have any questions.

**Comment by lukedemi** [comment]
@haotianw465 any update on lambda support?

**Comment by ss2305** [comment]
@lukedemiI It is still in our roadmap and is being prioritized against other features. Thanks for your continued enthusiasm with XRay.

**Comment by bonybrown** [comment]
Again, any update?

In the meantime, for those looking for a quick solution, this lambda function template should work...

```ruby
require 'aws-xray-sdk'

##
# START OF AWS XRAY SETUP
##

module XRay
  class LambdaStreamer < XRay::DefaultStreamer
    def initialize
      @stream_threshold = 1 #Stream every subsegment as it is available
    end
  end
  
  class LambdaEmitter < XRay::DefaultEmitter
    def send_entity(entity:)
      return nil if entity.name == '127.0.0.1' #Do not send localhost entities. It's the ruby lambda runtime
      super
    end
  end
  
  class LambdaContext < XRay::DefaultContext
    def handle_context_missing
      nil
    end

    def check_context
      #Create a new FacadeSegment if the _X_AMZN_TRACE_ID changes. 
      if ENV['_X_AMZN_TRACE_ID'] != @current_trace_id
        #puts "XRAY: Starting new segment for #{ENV['_X_AMZN_TRACE_ID']}"
        @current_trace_id = ENV['_X_AMZN_TRACE_ID']
        trace_header = XRay::TraceHeader.from_header_string(header_str: @current_trace_id)
        segment = FacadeSegment.new(trace_id: trace_header.root,
          parent_id: trace_header.parent_id,
          id: trace_header.parent_id,
          name: 'lambda_context',
          sampled: trace_header.sampled
        )
        store_entity(entity: segment)
      end
    end

    def current_entity
      check_context #ensure the FacadeSegment is current whenever the current_entity is retrieved
      super
    end
  end

  class FacadeSegment < XRay::Segment
    def initialize(trace_id: nil, name: nil, parent_id: nil, id: nil, sampled: true)
      super(trace_id: trace_id, name: name, parent_id: parent_id)
      @id = id
      @sampled = sampled
    end

    def ready_to_send?
      false #never send this facade. AWS Lambda has already created a Segment with these ids
    end
  end
end

config = {
  patch: %I[net_http aws_sdk],
  context: XRay::LambdaContext.new,
  streamer: XRay::LambdaStreamer.new,
  emitter: XRay::LambdaEmitter.new
}
XRay.recorder.configure(config)

##
# END OF AWS XRAY SETUP
##


def handler(event:,context:)
  # lambda implementation code goes here.
  # Every AWS SDK call and HTTP request will be instrumented
end
```

**Comment by ss2305** [comment]
@bonybrown Hi Tony. 
Thanks for posting the work around.This is still on our roadmap and will be picked up as soon as feasible. Unfortunately, I cannot provide an ETA at the moment. 
Thanks for posting the work around. Meanwhile, you are welcome to post a PR and we would more than happy to take a look.

**Comment by srprash** [comment]
Lambda support has been released in v0.11.4
